### DISCUSSION TITLE
Docker Pull Rate Limits - how to use a 'registry' proxy?

### DISCUSSION THREAD

**Discussion by maltemorgenstern** [renovatebot/renovate#18053] [Request Help] [answered]
Hey everyone,
thanks for maintaining renovate - it is a fantastic tool 🚀

At the moment we are experiencing some issues with Docker Pull Rate Limits. As far as I can tell we are already using every configuration option available to solve this - that's why I am opening this discussion.

### Our setup
We are running the `renovate/renovate` container inside GitLab CI/CD in our self-hosted GitLab CE instance. Renovate runs as a scheduled pipeline every hour and updates ~750 repositories.

Our teams are using public images from `docker.io` and `gcr.io` but also from an internal Harbor instance `harbor.company.local`.

For those docker registries we configured global `hostRules` for authentication:
```
hostRules: [
    {
      hostType: "docker",
      matchHost: "docker.io",
      username: process.env.DOCKER_IO_USERNAME,
      password: process.env.DOCKER_IO_PASSWORD,
      timeout: 30000
    },
    {
      hostType: "docker",
      matchHost: "harbor.company.local",
      username: process.env.HARBOR_USERNAME,
      password: process.env.HARBOR_PASSWORD,
      timeout: 30000
    }
]
```

We are using redis-based caching via a shared redis instance to reduce external requests:
```
repositoryCache: "enabled",
redisUrl: process.env.REDIS_URL
```

### Steps we already took
In order to configure authentication for `docker.io` and reduce requests we took the following steps:
- `hostRules` are configured for `docker.io` (see "Our setup") using a service account
- We run `docker login` inside the `renovate/renovate` container
  ```
  before_script:
    # Docker login (pull rate limit)
    - echo "${DOCKER_IO_PASSWORD}" | docker login --username "${DOCKER_IO_USERNAME}" --password-stdin
  script:
    - renovate ${RENOVATE_EXTRA_FLAGS} ${RENOVATE_REPOSITORIES}
  ```
- We are using Renovate v32.209.0 which already contains the `HEAD instead of GET` optimizations released in v26.14.0: https://github.com/renovatebot/renovate/issues/11473
- We are using `dockerImagePrefix` to pull the sidecar containers via `registry` proxy running at `registry.company.local`: https://docs.renovatebot.com/self-hosted-configuration/#dockerimageprefix
  ```
  # config.js
  dockerImagePrefix: "registry.company.local/renovate"
  ```
- We are actively monitoring the remaining pull rate limit for our service account (we followed https://about.gitlab.com/blog/2020/11/18/docker-hub-rate-limit-monitoring/) - which does not indicate any problems

### Using `registry` as proxy for all docker requests
As another improvement I was thinking about using our `registry` instance as a proxy for all requests to `docker.io`.

This idea is based on these issues/comments:
- https://github.com/renovatebot/renovate/issues/16040
- https://github.com/renovatebot/renovate/issues/10135#issuecomment-1013111534
- https://github.com/renovatebot/renovate/discussions/13799#discussioncomment-2049774

When adding the global `packageRules` it seems to work fine for updating versions - but not for updating digests:
```
"packageRules": [
  {
    "matchDatasources": ["docker"],
    "registryUrls": [
      "https://registry.company.local"
    ]
  }
] 
```

Looking for manifests does not use the `library` fallback which seems to result in HTTP 500 errors inside the `registry` container.
```
"message": "Response code 500 (Internal Server Error)",
"stack": "HTTPError: Response code 500 (Internal Server Error)\n    at Request.<anonymous> (/workspaces/renovate/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
"options": {
  "headers": {
    "user-agent": "RenovateBot/0.0.0-semantic-release (https://github.com/renovatebot/renovate)",
    "accept": "application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.docker.distribution.manifest.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json",
    "accept-encoding": "gzip, deflate, br"
  },
  "url": "https://registry.company.local/v2/alpine/manifests/sha256:686d8c9dfa6f3ccfc8230bc3178d23f84eeaf7e457f36f271ab1acc53015037c",
  "hostType": "docker",
  "username": "",
  "password": "",
  "method": "HEAD",
  "http2": false
},
"response": {
  "statusCode": 500,
  "statusMessage": "Internal Server Error",
  "body": "",
  "headers": {
    "content-type": "application/json; charset=utf-8",
    "docker-distribution-api-version": "registry/2.0",
    "x-content-type-options": "nosniff",
    "date": "Fri, 30 Sep 2022 08:11:20 GMT",
    "content-length": "289",
    "connection": "close"
  },
  "httpVersion": "1.1"
}
```

### Conclusion
As far as I can tell we are using every configuration option to solve the Docker Pull Rate Limit errors. Using a `registry` proxy as another solution does not seem to work.

Do you have any ideas/suggestions to solve this problem? Did I miss any setting?

Regards,
Malte

**Comment by viceice** [comment]
harbor isn't compatible as pull through proxy for renovate, it it doesn't list all upstream tags. 

harbor only returns already pulled docker tags.

a better proxy is the official `registry` docker image.

**Reply by maltemorgenstern** [reply]
Hey @viceice,
thanks for the quick response.

We are not using harbor as a pull through proxy.

We have two servers:
- `harbor.company.local`
  This server is running harbor as an internal registry for images created by us. We push and pull them from there.
- `registry.company.local`
  This server is running the official `registry` container (v2.8.1) as a pull through proxy for all our CI/CD servers.

This proxy is working great for the sidecar containers (`dockerImagePrefix`) - but not for updating hashes in other images.

If I can provide any additional information please tell me.

Regards,
Malte

**Reply by viceice** [reply]
renovate can't use a docker proxy by default. maybe you can workaround with `registryAliases` config option.

**Reply by maltemorgenstern** [reply]
Hey @viceice,
okay thanks for the info.

I will take a look at `registryAliases` - maybe we can configure that for `docker.io` dependencies 🤔

Do you know any additional settings/options we could/should use when it comes to `docker.io` authentication and pull rate limits?

Regards,
Malte

**Reply by viceice** [reply]
you should use a dedicated account. most renovate requests are not counted. only when the full docker manifest is fetched then it counts to limit.

but official images and some others are also excluded from limits.

**Reply by maltemorgenstern** [reply]
Mhh, maybe we actually have too many requests going to `docker.io` 🤔
Even after creating a new, dedicated docker account tens of projects still run into `TOOMANYREQUESTS` errors.

Using `registryAliases` without configuring any prefix/alias did not work either:
```
### Dockerfile
FROM renovate/renovate:32.216.0-slim

### config.js
registryAliases: {
  "": "registry.company.local"
}
```
This results in malformed lookup URLs 🤷‍♂️

**Reply by maltemorgenstern** [reply]
@viceice do you think a fallback to `library/` as added in https://github.com/renovatebot/renovate/pull/16509 could be implemented for `getManifestResponse` aswell?

https://github.com/renovatebot/renovate/blob/ea0eb67b108920ed5519041f83ec3139185ed8b6/lib/modules/datasource/docker/index.ts#L896-L904

This way one might be able to use a `registry` container as a proxy for all requests going to `docker.io` without changing any URLs inside Dockerfiles etc.

It would require some changes around here:
https://github.com/renovatebot/renovate/blob/ea0eb67b108920ed5519041f83ec3139185ed8b6/lib/modules/datasource/docker/index.ts#L448-L450

Regards,
Malte

**Reply by viceice** [reply]
I'm not sure if it will work as expected. it should already fallback.

**Reply by maltemorgenstern** [reply]
I did some more testing and created a new issue:
- https://github.com/renovatebot/renovate/issues/18206 

I think there are just a few cases where a fallback needs to be added to renovate in order to get this `registry` proxy up and running.

I will close this discussion.
Thanks a lot @viceice four your ideas/thoughts 🥇

**Comment by maltemorgenstern** [comment, marked as answer]
Closed in favor of
- https://github.com/renovatebot/renovate/issues/18206
